specs/core/0_beacon-chain.md

```diff
@@ -1117,26 +1119,35 @@ def slash_validator(state: BeaconState,
 
 ### Genesis trigger
 
-Before genesis has been triggered and whenever the deposit contract emits a `Deposit` log, call the function `is_genesis_trigger(deposits: Sequence[Deposit], timestamp: uint64) -> bool` where:
+Before genesis has been triggered and for every Eth 1.0 block call `is_genesis_trigger(deposits, time)` where:
 
-* `deposits` is the list of all deposits, ordered chronologically, up to and including the deposit triggering the latest `Deposit` log
-* `timestamp` is the Unix timestamp in the Ethereum 1.0 block that emitted the latest `Deposit` log
+* `deposits` is the list of all deposits up to the Eth 1.0 block, ordered chronologically
+* `time` is the Unix time of the Eth 1.0 block
 
-When `is_genesis_trigger(deposits, timestamp) is True` for the first time, let:
+When `is_genesis_trigger(deposits, time) is True` for the first time let:
 
 * `genesis_deposits = deposits`
-* `genesis_time = timestamp - timestamp % SECONDS_PER_DAY + 2 * SECONDS_PER_DAY` where `SECONDS_PER_DAY = 86400`
+* `genesis_time = time - time % SECONDS_PER_DAY + 2 * SECONDS_PER_DAY` where `SECONDS_PER_DAY = 86400`
 * `genesis_eth1_data` be the object of type `Eth1Data` where:
-    * `genesis_eth1_data.block_hash` is the Ethereum 1.0 block hash that emitted the log for the last deposit in `deposits`
     * `genesis_eth1_data.deposit_root` is the deposit root for the last deposit in `deposits`
     * `genesis_eth1_data.deposit_count = len(genesis_deposits)`
+    * `genesis_eth1_data.block_hash` is the Eth 1.0 block hash that emitted the log for the last deposit in `deposits`
 
 *Note*: The function `is_genesis_trigger` has yet to be agreed upon by the community, and can be updated as necessary. We define the following testing placeholder:
 
 ```python
-def is_genesis_trigger(deposits: Sequence[Deposit], timestamp: uint64) -> bool:
-    # Process deposits
+def is_genesis_trigger(deposits: List[Deposit, 2**DEPOSIT_CONTRACT_TREE_DEPTH], time: uint64) -> bool:
+    SECONDS_PER_DAY = 86400
+    # Do not deploy too early
+    if time - time % SECONDS_PER_DAY + 2 * SECONDS_PER_DAY < MIN_GENESIS_TIME:
+        return False
+
+    # Initialize deposit root
     state = BeaconState()
+    state.eth1_data.deposit_root = hash_tree_root(
+        Vector[DepositData, len(deposits)](list(map(lambda deposit: deposit.data, deposits)))
```

Comment:
Ahhhh, in the telegram chat discussion, I said that "I just understand that the `Deposit.proof` you mean is targeting on the `deposit_root` after the given DepositEvent is triggered… 🙈 (I was thinking to regenerate the proof with the latest deposit_root when triggering the genesis as what the dev branch is, but that’s much more overhead)" but it turns out you *did* mean using the same `deposit_root` for all the `proof`s! 😅 

We expect the implementers would optimize it with their own approach. However, this function is part of the consensus tests and the implementers can use this function to testify the optimized version.

Generating `Deposit.proof` with the same `deposit_root` disallows the implementation to build the tree `DepositData` by `DepositData`. So I now think using `deposit_root` per `DepositEvent` might be more realistic and providing the same testing input as the clients would do.

/cc @terencechain since I think Prysm already has the deposit contract handler? Any thought? :)

Reply:
I don't mind either way. Changed it to incrementally compute `deposit_root`.